classes/announcement/DAO.inc.php

```diff
@@ -0,0 +1,133 @@
+<?php
+/**
+ * @file classes/announcement/DAO.inc.php
+ *
+ * Copyright (c) 2014-2021 Simon Fraser University
+ * Copyright (c) 2000-2021 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class announcement
+ *
+ * @brief Read and write announcements to the database.
+ */
+
+namespace PKP\announcement;
+
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\App;
+use Illuminate\Support\LazyCollection;
+use PKP\core\EntityDAO;
+use stdClass;
+
+class DAO extends EntityDAO
+{
+    /** @copydoc EntityDAO::$schema */
+    public string $schema = \PKP\Services\PKPSchemaService::SCHEMA_ANNOUNCEMENT;
+
+    /** @copydoc EntityDAO::$table */
+    public string $table = 'announcements';
+
+    /** @copydoc EntityDAO::$settingsTable */
+    public string $settingsTable = 'announcement_settings';
+
+    /** @copydoc EntityDAO::$primarykeyColumn */
+    public string $primaryKeyColumn = 'announcement_id';
+
+    /** @copydoc EntityDAO::$primaryTableColumns */
+    public array $primaryTableColumns = [
+        'id' => 'announcement_id',
+        'assocId' => 'assoc_id',
+        'assocType' => 'assoc_type',
+        'typeId' => 'type_id',
+        'dateExpire' => 'date_expire',
+        'datePosted' => 'date_posted',
+    ];
+
+    /**
+     * @copydoc EntityDAO::newDataObject()
+     */
+    public function newDataObject(): Announcement
+    {
+        return App::make(Announcement::class);
+    }
+
+    /**
+     * @copydoc EntityDAO::get()
+     */
+    public function get(int $id): Announcement
```

Comment:
I still miss the old convention where we could include (where it makes sense) an optional second parameter for e.g. a context ID. This makes it easy to add built-in validation that the `Announcement` is in the expected `Context`, which is easily overlooked. (I suppose that would be best offered to downstream code in `Query`?)

Reply:
My concern with this is that it obscures the reason that an object was not found. When writing a response for a request, especially in the API, it is nice to be able to differentiate between "this item doesn't exist" and "this item is in another context".

We don't use this a lot, yet, but we do use it some in the REST API and I can see it being used in CLIs too, where the user is not always clear what context they're working in.

My feeling is that the DAO should be as simple/dumb as possible, and that things like authorization/validation should occur within the `Handler`, which is the only place that understands the request.

I'm willing to budge on this, but I think we'd also have to consider every object separately. Announcements and Submissions can receive a Context ID, but ReviewAssignments and Publications probably ought to get a Submission ID instead. And Galleys and Authors ought to get a Publication ID. And SubmissionFiles ought to get... well that would be tricky.

Reply:
No, the more I think about it, the more we were just compensating for not having a standard place to do proper checks. I think this will be less of an issue as we adopt the API for internal use. Carry on!